Sources/StreamChat/WebSocketClient/EventMiddlewares/ChannelReadUpdaterMiddleware.swift

```diff
@@ -50,36 +56,124 @@ struct ChannelReadUpdaterMiddleware: EventMiddleware {
+                subsystems: .webSocket
+            )
         }
-
-        // Try to get the existing channel read for the current user
-        if let read = session.loadChannelRead(cid: cid, userId: currentUserId) {
-            if message.createdAt > read.lastReadAt ?? Date.distantPast {
-                read.unreadMessageCount += 1
-            }
-        } else {
-            log.info(
-                "Can't increase unread count for \(cid) because the `ChannelReadDTO` for " +
-                    "the current user doesn't exist."
+        
+        log.debug(
+            "Message \(message.id) increments unread counts for channel \(cid)",
+            subsystems: .webSocket
+        )
+        
+        channelRead.unreadMessageCount += 1
+    }
+    
+    private func decrementUnreadCountIfNeeded(
+        event: MessageDeletedEventDTO,
+        session: DatabaseSession
+    ) {
+        guard let currentUser = session.currentUser else {
+            return log.error("Current user is missing", subsystems: .webSocket)
+        }
+        
+        guard let channelRead = session.loadChannelRead(cid: event.cid, userId: currentUser.user.id) else {
+            return log.error("Channel read is missing", subsystems: .webSocket)
+        }
+        
+        if let skipReason = !event.hardDelete
+            ? .messageIsSoftDeleted
+            : unreadCountUpdateSkippingReason(
+                currentUser: currentUser,
+                channelRead: channelRead,
+                message: event.message
+            ) {
+            return log.debug(
+                "Message \(event.message.id) does not decrement unread coutns for \(event.message.user.id): \(skipReason)",
+                subsystems: .webSocket
             )
         }
+        
+        log.debug(
+            "Message \(event.message.id) decrements unread counts for channel \(event.cid)",
+            subsystems: .webSocket
+        )
+        
+        channelRead.unreadMessageCount = max(0, channelRead.unreadMessageCount - 1)
+    }
+        
+    private func unreadCountUpdateSkippingReason(
+        currentUser: CurrentUserDTO,
+        channelRead: ChannelReadDTO,
+        message: MessagePayload
+    ) -> UnreadSkippingReason? {
+        if channelRead.channel.mute != nil {
+            return .channelIsMuted
+        } else if message.user.id == currentUser.user.id {
+            return .messageIsOwn
+        } else if currentUser.mutedUsers.contains(where: { message.user.id == $0.id }) {
+            return .authorIsMuted
+        } else if message.isSilent {
+            return .messageIsSilent
+        } else if message.parentId != nil && !message.showReplyInChannel {
+            return .messageIsThreadReply
+        } else if message.type == .system {
+            return .messageIsSystem
+        } else if message.createdAt <= (channelRead.lastReadAt ?? Date.distantPast) {
+            return .messageIsSeen
+        } else {
+            return nil
+        }
```

Comment:
nit: The `else if` are not necessary because we use returns, we could do just:
```swift
if ... {
    return 
}

if ... {
    return 
}

if ... {
    return 
}

....
```